tests/chainer_tests/utils_tests/test_conv_nd.py

```diff
@@ -0,0 +1,315 @@
+        expected = (1, 2, 3)
+        self.assertEqual(actual, expected)
+
+    def test_tuple_invalid_length(self):
+        with self.assertRaises(AssertionError):
+            conv_nd.as_tuple((1,), 3)
+
+
+@testing.parameterize(*testing.product({
+    'dims': [(10,), (10, 8), (10, 8, 6)],
+}))
+class TestIm2ColND(unittest.TestCase):
+
+    def setUp(self):
+        shape = (2, 3) + self.dims
+        self.img = numpy.random.uniform(-1, 1, shape).astype(numpy.float32)
+
+    def check_im2col_nd(self, ksize, stride, pad, gpu):
+        dims = self.dims
+        if gpu:
+            im2col = conv_nd.im2col_nd_gpu
+            img = cuda.to_gpu(self.img)
+        else:
+            im2col = conv_nd.im2col_nd_cpu
+            img = self.img
+
+        col = im2col(img, ksize, stride, pad)
+        outs = tuple(conv_nd.get_conv_outsize(d, k, s, p)
+                     for (d, k, s, p) in zip(dims, ksize, stride, pad))
+        expected_shape = (2, 3) + ksize + outs
+        self.assertEqual(col.shape, expected_shape)
+
+        col = cuda.to_cpu(col)
+
+        for n in moves.range(2):
+            for c in moves.range(3):
+                for xs in itertools.product(
+                        *[moves.range(out) for out in outs]):
+                    for dxs in itertools.product(
+                            *[moves.range(k) for k in ksize]):
+                        oxs = tuple(x * s - p + dx
+                                    for (x, s, p, dx)
+                                    in zip(xs, stride, pad, dxs))
+                        if all(0 <= ox < d for (ox, d) in zip(oxs, dims)):
+                            col_index = (n, c) + dxs + xs
+                            img_index = (n, c) + oxs
+                            self.assertEqual(
+                                col[col_index], self.img[img_index])
+                        else:
+                            col_index = (n, c) + dxs + xs
+                            self.assertEqual(col[col_index], 0)
+
+    def test_im2col_nd_1_cpu(self):
+        ndim = len(self.dims)
+        ksize = (1,) * ndim
+        stride = (1,) * ndim
+        pad = (1,) * ndim
+        self.check_im2col_nd(ksize, stride, pad, gpu=False)
+
+    def test_im2col_nd_2_cpu(self):
+        ndim = len(self.dims)
+        ksize = (2,) * ndim
+        stride = (2,) * ndim
+        pad = (2,) * ndim
+        self.check_im2col_nd(ksize, stride, pad, gpu=False)
+
+    def test_im2col_nd_3_cpu(self):
+        ndim = len(self.dims)
+        ksize = (1, 2, 1)[:ndim]
+        stride = (2, 1, 2)[:ndim]
+        pad = (1, 2, 1)[:ndim]
+        self.check_im2col_nd(ksize, stride, pad, gpu=False)
+
+    @attr.gpu
+    def test_im2col_nd_1_gpu(self):
+        ndim = len(self.dims)
+        ksize = (1,) * ndim
+        stride = (1,) * ndim
+        pad = (1,) * ndim
+        self.check_im2col_nd(ksize, stride, pad, gpu=True)
+        self.assertTrue(ndim in conv_nd._im2col_cache)
```

Comment:
Should we ensure `_im2col_cache` being empty before we execute `im2col_nd_gpu` ?